Revert "Lambert and _is_lambert modifications"

oscarbenjamin opened the pull request:
Reverts sympy/sympy#20613

See discussion in the PR. I think I merged it by accident somehow and there are problems with it.

@MaqOwais I think that #20613 had some good work in it but there are some basic problems with it as well. I have made some suggestions but it seems that you do not have a clear idea how to fix this because #20705 has gone off on a massive tangent with 500 lines changed and adding new arguments to `solve`. Changes to `solve` are not what is needed since this is about `solveset` and `solveset` should not use `solve`.

I think that the best thing now is to revert #20613 and then start again with this. Note that the `solveset` code should not use `solve`. Also if the basic functions for Lambert forms you are using are from `solve` and return incomplete solution sets then we do not want to use those as part of `solveset` either. It is very important that `solveset` only ever returns complete solutions or otherwise returns `ConditionSet`.

I think you now have a better idea what the full solution sets should look like for solutions involving LambertW but I think that we need to start from a clean implementation so it's best to revert #20613 for now and focus on getting a new implementation that is correct from the start even if it only handles e.g. the real case.

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

oscarbenjamin committed:
Revert "Lambert and _is_lambert modifications"

MaqOwais commented:
> Note that the solveset code should not use solve...

Already solve is used in bivariate.py which is under lambert and solveset...

https://github.com/sympy/sympy/blob/841ce6b95dbd398c195472aaccb018012869ef8b/sympy/solvers/bivariate.py#L151-L160

oscarbenjamin commented:
Maybe it is fine to use those functions from `_lambert` but only if the domain is a subset of the reals. We should study carefully each case that is passed to it though and ensure that the solution set is entirely complete. It is better to add cases carefully one at a time making sure that all of the results are correct.

I think it would be better if `solveset` did not use `solve` at all but it is probably okay in certain limited situations e.g. if the equation is known to be invertible. I think that `solveset` has better functions for those cases though.

MaqOwais commented:
> Maybe it is fine to use those functions from _lambert but only if the domain is a subset of the reals.

But in such case, the domain consistency will not be retained...
If a `solve` function is called form `_lambert` then the domain will be changed to complexes(from `S.Reals` if the domain is S.Reals) since other helper functions will tend to make `S.Complexes` as a default value.

So to overcome such an issue either we have to consider using global functions or passing the arguments all over the solve.py functions ( as done in #20705 ) ... or should think about any other way to maintain consistency...

oscarbenjamin commented:
If the `solve` functions can not be restricted to the reals and they also don't have a way to get the full solution over the complexes then they should not be used in solveset. Instead new functions will need to be made that have clear and well-defined behaviour giving complete solutions over some domain.

MaqOwais commented:
> If the solve functions can not be restricted to the reals and they also don't have a way to get the full solution over the complexes then they should not be used in solveset.

But until now `solve`(under `solveset` <particularly in lambert cases>) is giving the required/complete solutions as until now atleast in case of lambert functions.

oscarbenjamin commented:
I don't know what you mean by "until now". Do you mean before I reverted #20613?

Now on current master `solveset` does not work with Lambert equations just like before this PR was merged:
```python
In [1]: solveset(x*exp(x)-2)
Out[1]: 
⎧            ⎛   x        ⎞⎫
⎨x | x ∊ ℂ ∧ ⎝x⋅ℯ  - 2 = 0⎠⎬
⎩                          ⎭
```
I reverted #20613 because there were cases where an incomplete solution set was returned. I have already listed many examples of this.

MaqOwais commented:
> I don't know what you mean by "until now". Do you mean before I reverted #20613?

yes

MaqOwais commented:
```
>>> solveset(x*exp(x)-2,x,S.Complexes)
FiniteSet(LambertW(2))
>>> solveset(x*exp(x)-2,x,S.Reals)
FiniteSet(LambertW(2))
```

oscarbenjamin commented:
> I think `solveset` in S.Complexes should have returned ` Imageset(Lambda(n, LambertW(2,n)))`

Exactly. The solution set was not complete.

The reason for creating the `solveset` function as a replacement for `solve` was because the behaviour of `solve` is underspecified and it is never possible to know if any results returned from `solve` represent a complete solution set (actually they often don't because `solve` does not represent infinite solution sets).

The whole point of `solveset` is to have the clearly defined behaviour that any result it returns will be complete. A ConditionSet is a way of doing this without returning an explicit solution. The ConditionSet allows us to return something that is not necessarily the empty set in the cases where we are unable to get an explicit representation of the *complete* set of solutions.

MaqOwais commented:
> I think solveset in S.Complexes should have returned Imageset(Lambda(n, LambertW(2,n)))
> Actually I forget to add this case b.t.w...

I think this could be achieved by simply adding one more case, in #20613

oscarbenjamin commented:
> I think this could be achieved by simply adding one more case, in #20705

That PR adds a `domain` argument to `solve` and we do not want to change the signature of `solve`. In `solve` the domain is specified by assumptions on the symbol:
```python
In [2]: x = Symbol('x')

In [3]: solve(x**3 - x**2 + x - 1, x)
Out[3]: [1, -ⅈ, ⅈ]

In [4]: x = Symbol('x', real=True)

In [5]: solve(x**3 - x**2 + x - 1, x)
Out[5]: [1]
```
We do not want to change `solve` to have a domain argument and we also do not want `solveset` to give output that depends on the assumptions of the unknown symbols.

MaqOwais commented:
Even in #20705, the scenario remains the same...

```
>>> x = Symbol('x')
>>> solve(x**3 - x**2 + x - 1, x)
[1, -I, I]
>>> x = Symbol('x', real=True)
>>> solve(x**3 - x**2 + x - 1, x)
[1]
```

Even though the scenario remains the same, I think it's not a good way to have domain=domain in solve.py...
Have to figure out another way to maintain consistency in the domain.

MaqOwais commented:
> (it only affects the lambert solutions while specifying domain)

E.g.
```
>>> solve(exp(x) - x, x)
[-LambertW(-1, integer)]
>>> solve(exp(x) - x, x,domain=S.Reals)
[]
```
